src/django_idom/hooks.py

```diff
@@ -41,84 +50,72 @@ def use_websocket() -> IdomWebsocket:
     return websocket
 
 
-_REFETCH_CALLBACKS: DefaultDict[FunctionType, set[Callable[[], None]]] = DefaultDict(
-    set
-)
+_REFETCH_CALLBACKS: DefaultDict[
+    Callable[..., Any], set[Callable[[], None]]
+] = DefaultDict(set)
 
 
-_Data = TypeVar("_Data")
+_Result = TypeVar("_Result", bound=Union[Model, QuerySet[Any]])
 _Params = ParamSpec("_Params")
 
 
 def use_query(
-    query: Callable[_Params, _Data],
+    query: Callable[_Params, _Result | None],
     *args: _Params.args,
-    fetch_deferred_fields: bool = True,
     **kwargs: _Params.kwargs,
-) -> Query[_Data]:
-    given_query = query
-    query, _ = use_state(given_query)  # type: ignore
-    if given_query is not query:
-        raise ValueError(f"Query function changed from {query} to {given_query}.")
+) -> Query[_Result | None]:
+    query_ref = use_ref(query)
+    if query_ref.current is not query:
+        raise ValueError(f"Query function changed from {query_ref.current} to {query}.")
 
-    data, set_data = use_state(UNDEFINED)
+    should_execute, set_should_execute = use_state(True)
+    data, set_data = use_state(cast(Union[_Result, None], None))
     loading, set_loading = use_state(True)
-    error, set_error = use_state(None)
+    error, set_error = use_state(cast(Union[Exception, None], None))
 
     @use_callback
     def refetch() -> None:
-        set_data(UNDEFINED)
+        set_should_execute(True)
+        set_data(None)
         set_loading(True)
         set_error(None)
 
     @use_effect(dependencies=[])
-    def add_refetch_callback():
+    def add_refetch_callback() -> Callable[[], None]:
         # By tracking callbacks globally, any usage of the query function will be re-run
         # if the user has told a mutation to refetch it.
-        _REFETCH_CALLBACKS[query].add(refetch)  # type: ignore
-        return lambda: _REFETCH_CALLBACKS[query].remove(refetch)  # type: ignore
+        _REFETCH_CALLBACKS[query].add(refetch)
+        return lambda: _REFETCH_CALLBACKS[query].remove(refetch)
 
     @use_effect(dependencies=None)
-    def execute_query():
-        if data is not UNDEFINED:
+    def execute_query() -> None:
+        if not should_execute:
             return
 
-        def thread_target():
+        def thread_target() -> None:
             # sourcery skip: remove-empty-nested-block, remove-redundant-pass
             try:
                 query_result = query(*args, **kwargs)
             except Exception as e:
-                set_data(UNDEFINED)
+                set_data(None)
                 set_loading(False)
-                set_error(e)  # type: ignore
+                set_error(e)
                 return
+            finally:
+                set_should_execute(False)
 
-            if isinstance(query_result, QuerySet):  # type: ignore
-                if fetch_deferred_fields:
-                    for model in query_result:
-                        _fetch_deferred_fields(model)
```

Comment:
Is there a reason why this was removed?

Reply:
my bad, refactored anticipating a section function and forgot to put this back. Will fix that tomorrow.